traitlets/utils/bunch.py

```diff
@@ -19,7 +41,42 @@ def __setattr__(self, key, value):
     
     def __dir__(self):
         # py2-compat: can't use super because dict doesn't have __dir__
-        names = dir({})
+        names = dir(type(self))
         names.extend(self.keys())
         return names
 
+
+class MetaFrozenBunch(type):
+    
+    def __new__(mcls, name, bases, classdict):
+        for n in ('clear', 'pop', 'popitem', 'setdefault', 'update'):
+            classdict[n] = EmptyMethodDescriptor(n, AttributeError, 'has no attribute %s' % n)
+        classdict['__setattr__'] = EmptyMethodDescriptor( '__setattr__',
+                        TypeError, 'does not support attribute assignment')
+        classdict['__delattr__'] = EmptyMethodDescriptor( '__delattr__',
+                        TypeError, 'does not support attribute deletion')
+        classdict['__setitem__'] = EmptyMethodDescriptor( '__setitem__',
+                        TypeError, 'does not support item assignment')
+        classdict['__delitem__'] = EmptyMethodDescriptor('__delitem__',
+                        TypeError, 'does not support item deletion')
+        return super(MetaFrozenBunch, mcls).__new__(mcls, name, bases, classdict)
+
+
+class FrozenBunch(with_metaclass(MetaFrozenBunch, Bunch)):
```

Comment:
@minrk, are you saying that we don't need to freeze the bunch, or that there must be a better way to freeze it.